File: Java_version/app/src/main/java/com/termux/wayland/scanner/JsonAppScanner.java
package com.termux.wayland.scanner;

import android.content.Context;

import com.termux.wayland.data.LinuxApp;
import com.termux.wayland.data.LinuxApp.AppSource;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.File;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Lit les deux sources JSON :
 *  1. assets/apps.json      → base curatée bundlée dans l'APK
 *  2. apps_user.json        → fichier éditable par l'utilisateur
 *                             stocké dans filesDir de l'app
 *
 * Règle de merge : apps_user.json peut OVERRIDER ou AJOUTER des entrées.
 * Priorité : USER > DATABASE
 */
public class JsonAppScanner {

    private static final String USER_FILE_NAME = "apps_user.json";

    // Singleton
    private static JsonAppScanner instance;

    private JsonAppScanner() {}

    public static synchronized JsonAppScanner getInstance() {
        if (instance == null) {
            instance = new JsonAppScanner();
        }
        return instance;
    }

    /**
     * Retourne la liste brute mergée des deux JSON.
     * Non filtrée par dpkg.
     */
    public List<LinuxApp> loadApps(Context context) {
        List<LinuxApp> dbApps   = loadDatabase(context);
        List<LinuxApp> userApps = loadUserFile(context);
        return merge(dbApps, userApps);
    }

    // ─── Lecture assets/apps.json ─────────────────────────────────────────────

    private List<LinuxApp> loadDatabase(Context context) {
        try {
            InputStream is = context.getAssets().open("apps.json");
            byte[] buffer = new byte[is.available()];
            is.read(buffer);
            is.close();
            String json = new String(buffer, StandardCharsets.UTF_8);
            return parseJsonArray(json, AppSource.DATABASE);
        } catch (Exception e) {
            e.printStackTrace();
            return new ArrayList<>();
        }
    }

    // ─── Lecture apps_user.json (filesDir) ───────────────────────────────────

    private List<LinuxApp> loadUserFile(Context context) {
        File file = new File(context.getFilesDir(), USER_FILE_NAME);
        if (!file.exists()) return new ArrayList<>();
        try {
            byte[] bytes = new byte[(int) file.length()];
            java.io.FileInputStream fis = new java.io.FileInputStream(file);
            fis.read(bytes);
            fis.close();
            String json = new String(bytes, StandardCharsets.UTF_8);
            return parseJsonArray(json, AppSource.USER);
        } catch (Exception e) {
            e.printStackTrace();
            return new ArrayList<>();
        }
    }

    // ─── Parser JSON → List<LinuxApp> ────────────────────────────────────────

    private List<LinuxApp> parseJsonArray(String json, AppSource source) {
        List<LinuxApp> apps = new ArrayList<>();
        try {
            JSONArray array = new JSONArray(json);
            for (int i = 0; i < array.length(); i++) {
                JSONObject obj = array.getJSONObject(i);
                LinuxApp app = parseApp(obj, source);
                if (app != null) apps.add(app);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return apps;
    }

    private LinuxApp parseApp(JSONObject obj, AppSource source) {
        try {
            return new LinuxApp(
                obj.getString("id"),
                obj.getString("name"),
                obj.getString("package"),
                obj.getString("exec"),
                obj.optString("category", "tools"),
                obj.has("icon") ? obj.getString("icon") : null,
                obj.optBoolean("requires_display", true),
                obj.optString("repo", "x11-repo"),
                obj.optString("description", ""),
                source
            );
        } catch (Exception e) {
            return null; // entrée malformée → ignorée
        }
    }

    // ─── Merge : USER écrase DATABASE si même id ─────────────────────────────

    private List<LinuxApp> merge(List<LinuxApp> dbApps, List<LinuxApp> userApps) {
        // Index de la base par id (LinkedHashMap garde l'ordre d'insertion)
        Map<String, LinuxApp> merged = new LinkedHashMap<>();
        for (LinuxApp app : dbApps) {
            merged.put(app.getId(), app);
        }

        for (LinuxApp userApp : userApps) {
            LinuxApp existing = merged.get(userApp.getId());
            if (existing != null) {
                // Override partiel : on complète avec les valeurs de la base
                merged.put(userApp.getId(), new LinuxApp.Builder()
                    .from(existing)
                    .name(userApp.getName().isEmpty()        ? existing.getName()        : userApp.getName())
                    .packageName(userApp.getPackageName().isEmpty() ? existing.getPackageName() : userApp.getPackageName())
                    .exec(userApp.getExec().isEmpty()        ? existing.getExec()        : userApp.getExec())
                    .category(userApp.getCategory().isEmpty()? existing.getCategory()    : userApp.getCategory())
                    .icon(userApp.getIcon() != null          ? userApp.getIcon()         : existing.getIcon())
                    .requiresDisplay(userApp.isRequiresDisplay())
                    .repo(userApp.getRepo().isEmpty()        ? existing.getRepo()        : userApp.getRepo())
                    .description(userApp.getDescription().isEmpty() ? existing.getDescription() : userApp.getDescription())
                    .source(AppSource.USER)
                    .build()
                );
            } else {
                // Nouvelle entrée inconnue de la base
                merged.put(userApp.getId(), userApp);
            }
        }

        return new ArrayList<>(merged.values());
    }

    /**
     * Retourne le chemin du fichier apps_user.json.
     */
    public String getUserFilePath(Context context) {
        return new File(context.getFilesDir(), USER_FILE_NAME).getAbsolutePath();
    }

    /**
     * Crée un apps_user.json vide si inexistant.
     * À appeler au premier lancement.
     */
    public void initUserFileIfNeeded(Context context) {
        File file = new File(context.getFilesDir(), USER_FILE_NAME);
        if (!file.exists()) {
            try {
                java.io.FileWriter fw = new java.io.FileWriter(file);
                fw.write("[]");
                fw.close();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }
}

File: Kotlin_version/app/src/beta/java/com/termux/wayland/kotlin/beta/scanner/JsonAppScanner.kt
package com.termux.wayland.kotlin.beta.scanner

import android.content.Context
import com.termux.wayland.kotlin.beta.data.AppSource
import com.termux.wayland.kotlin.beta.data.LinuxApp
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

/**
 * Lit les deux sources JSON :
 *  1. assets/apps.json      → base curatée bundlée dans l'APK
 *  2. apps_user.json        → fichier éditable par l'utilisateur
 *                             stocké dans le dossier de données de l'app
 *
 * Règle de merge : apps_user.json peut OVERRIDER ou AJOUTER des entrées.
 * Un override est détecté par un "id" identique.
 * Priorité : USER > DATABASE
 */
object JsonAppScanner {

    // Chemin du fichier utilisateur dans le stockage interne de l'app
    private const val USER_FILE_NAME = "apps_user.json"

    /**
     * Retourne la liste brute des apps depuis les deux JSON,
     * mergée mais PAS encore filtrée par dpkg.
     *
     * @param context Android Context pour accéder aux assets
     */
    fun loadApps(context: Context): List<LinuxApp> {
        val dbApps = loadDatabase(context)
        val userApps = loadUserFile(context)
        return merge(dbApps, userApps)
    }

    // ─── Lecture apps.json (assets) ──────────────────────────────────────────

    private fun loadDatabase(context: Context): List<LinuxApp> {
        return try {
            val json = context.assets.open("apps.json")
                .bufferedReader()
                .readText()
            parseJsonArray(json, AppSource.DATABASE)
        } catch (e: Exception) {
            e.printStackTrace()
            emptyList()
        }
    }

    // ─── Lecture apps_user.json (fichier interne) ─────────────────────────────

    private fun loadUserFile(context: Context): List<LinuxApp> {
        val file = File(context.filesDir, USER_FILE_NAME)
        if (!file.exists()) return emptyList()
        return try {
            val json = file.readText()
            parseJsonArray(json, AppSource.USER)
        } catch (e: Exception) {
            e.printStackTrace()
            emptyList()
        }
    }

    // ─── Parser JSON → List<LinuxApp> ────────────────────────────────────────

    private fun parseJsonArray(json: String, source: AppSource): List<LinuxApp> {
        val apps = mutableListOf<LinuxApp>()
        try {
            val array = JSONArray(json)
            for (i in 0 until array.length()) {
                val obj = array.getJSONObject(i)
                val app = parseApp(obj, source) ?: continue
                apps.add(app)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return apps
    }

    private fun parseApp(obj: JSONObject, source: AppSource): LinuxApp? {
        return try {
            LinuxApp(
                id             = obj.getString("id"),
                name           = obj.getString("name"),
                packageName    = obj.getString("package"),
                exec           = obj.getString("exec"),
                category       = obj.optString("category", "tools"),
                icon           = obj.optString("icon", null),
                requiresDisplay= obj.optBoolean("requires_display", true),
                repo           = obj.optString("repo", "x11-repo"),
                description    = obj.optString("description", ""),
                source         = source
            )
        } catch (e: Exception) {
            null  // entrée malformée → on l'ignore silencieusement
        }
    }

    // ─── Merge : USER écrase DATABASE si même id ─────────────────────────────

    private fun merge(
        dbApps: List<LinuxApp>,
        userApps: List<LinuxApp>
    ): List<LinuxApp> {
        // Index de la base par id
        val merged = dbApps.associateBy { it.id }.toMutableMap()

        for (userApp in userApps) {
            val existing = merged[userApp.id]
            if (existing != null) {
                // Override partiel : on garde les champs non nuls de l'utilisateur
                // et on complète avec les valeurs de la base
                merged[userApp.id] = existing.copy(
                    name           = userApp.name.ifBlank { existing.name },
                    packageName    = userApp.packageName.ifBlank { existing.packageName },
                    exec           = userApp.exec.ifBlank { existing.exec },
                    category       = userApp.category.ifBlank { existing.category },
                    icon           = userApp.icon ?: existing.icon,
                    requiresDisplay= userApp.requiresDisplay,
                    repo           = userApp.repo.ifBlank { existing.repo },
                    description    = userApp.description.ifBlank { existing.description },
                    source         = AppSource.USER
                )
            } else {
                // Nouvelle entrée utilisateur inconnue de la base
                merged[userApp.id] = userApp
            }
        }

        return merged.values.toList()
    }

    /**
     * Retourne le chemin du fichier apps_user.json
     * pour permettre à l'UI de le montrer / l'ouvrir.
     */
    fun getUserFilePath(context: Context): String {
        return File(context.filesDir, USER_FILE_NAME).absolutePath
    }

    /**
     * Crée un apps_user.json vide si inexistant.
     * À appeler au premier lancement.
     */
    fun initUserFileIfNeeded(context: Context) {
        val file = File(context.filesDir, USER_FILE_NAME)
        if (!file.exists()) {
            file.writeText("[]")
        }
    }
}
